---
title: "TRAF_for_class"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("biomaRt")
install.packages("tidyverse")
install.packages(c("httr","jsonlite"))
install.packages("seqinr")

library(tidyverse)
library(biomaRt)
library(httr)
library(jsonlite)
library(seqinr)

#Get genes
IDs <- scan("diverse_proteomes.txt", character())
url_list <- tibble(ID = IDs,
                   URL = rep("https://www.ebi.ac.uk/interpro/api/protein/UniProt/entry/InterPro/IPR008974/proteome/uniprot/REPLACE/?page_size=300",length(IDs)),
                   URL2 = rep("https://www.ebi.ac.uk/interpro/api/protein/UniProt/entry/InterPro/IPR002083/proteome/uniprot/REPLACE/?page_size=300",length(IDs))
                   ) %>%
  mutate(URL = str_replace(URL, "REPLACE", ID)) %>%
  mutate(URL2 = str_replace(URL2, "REPLACE", ID))


res_ <- map(c(url_list$URL,url_list$URL2), GET)

saveRDS(res_, "res_.RDS")
res_<- readRDS("res_.RDS")

table(unlist(map(res_, c(status_code))))
# res_ <- discard(res_, ~.x$status_code == 204) #remove lists that don't have any data if any
data <- map(res_, "content") %>%
  map(~fromJSON(rawToChar(.))) %>%
  map(c("results","metadata","accession")) %>%
  unlist() %>%
  unique() # generates list of uniprot accessions containing interpro accessions - NOTE THAT SOME ISOFORMS MAY NOT BE INCLUDED

# IDs2 <- tibble(read_delim("tax_IDs_remainder.txt", col_names = FALSE, delim= "\n",col_types = "c"))
# url_list_ <- tibble(ID = unlist(IDs2),
#                    URL = rep("https://www.ebi.ac.uk/interpro/api/protein/UniProt/entry/InterPro/IPR008974/taxonomy/uniprot/REPLACE/?page_size=300",length(IDs2))) %>% 
#     mutate(URL = str_replace(URL, "REPLACE", ID))
# 
# res__ <- map(url_list_$URL, GET)
# res__ <- discard(res__, ~.x$status_code == 204)
# data2 <- map(res__, "content") %>%
#   map(~fromJSON(rawToChar(.))) %>%
#   map(c("results","metadata","accession")) %>%
#   purrr::flatten() %>%
#   as.character()%>%
#   as_tibble() %>%
#   distinct()
# 
# data_all <- rbind(data,data2)


url_list3 <- tibble(Accession = data, URL = rep("https://www.ebi.ac.uk/interpro/api/entry/InterPro/protein/UniProt/REPLACE?page_size=350",length(data))) %>%
  mutate(URL = str_replace(URL, "REPLACE", Accession)) 

res_3 <- map(url_list3$URL, GET) #get interpro data for each accession
table(unlist(map(res_3, c(status_code))))
# res_3 <- discard(res_3, ~.x$status_code == 204)

saveRDS(res_3, "res_3.RDS")
res_3<- readRDS("res_3.RDS")


meta <- map(res_3, "content") %>%
  map(~fromJSON(rawToChar(.))) %>%
  map("results")
  

itp_meta <- meta %>%
  map(c("metadata")) %>%
  map(~dplyr::select(.x, "accession","name","source_database","type")) %>%
  map(transpose)

itp_prot <- meta %>%
  map(c("proteins")) %>%
  map(~map(.x, ~dplyr::rename(.x, "uniprot_accession" = "accession"))) %>%
  map(~map(.x, ~dplyr::rename(.x, "uniprot_source_database" = "source_database"))) %>%
  map(~map(.x, purrr::flatten)) %>%
  map(~map(.x, purrr::flatten))

itp_ <-map2(itp_meta, itp_prot, ~map2(.x,.y, c))

itp_names <- itp_ %>%
  map(~map(., names)) %>%
  map(unlist) %>%
  purrr::reduce(union)

itp_ <- itp_ %>%
  map(transpose, .names = itp_names) %>%
  # map(~map_if(.x, is.data.frame, list)) %>%
  map(~as_tibble(.x, .name_repair = "unique")) %>%
  map(~unnest(.x, c("fragments", "model", "score"))) %>%
  map(~unnest(.x, fragments)) %>%
  map(transpose)

itp_names2 <- itp_ %>%
  map(~map(., names)) %>%
  map(unlist) %>%
  purrr::reduce(union) # to make sure all columns are kept

itp__ <- itp_ %>%
  map(transpose, .names = itp_names2) %>%
  map(~as_tibble(.x, .name_repair = "unique"))

pdl <- bind_rows(itp__, .id="Order") %>%
  unnest(cols = colnames(.)) %>%
  mutate(uniprot_accession = toupper(uniprot_accession))

# pdl_map <- pdl %>%
#   group_by(Order, uniprot_accession) %>%
#   summarise() %>%
#   drop_na(uniprot_accession)

pdl_type <- pdl %>%
  filter(type == "homologous_superfamily") %>%
  mutate(domain_code = recode(accession, "IPR008974" = "T", "IPR011333" = "B", "IPR013083" = "Z", "IPR038765" = "P", .default = "")) %>%
  mutate(mid = (end+start)/2) %>%
  group_by(uniprot_accession) %>%
  arrange(mid, .by_group = TRUE) %>%
  mutate(protein_type = paste0(domain_code, collapse = "")) %>%
  ungroup() %>%
  dplyr::select(uniprot_accession, protein_type) %>%
  distinct(uniprot_accession, .keep_all = TRUE)

pdl_type2 <- pdl %>%
  filter(type == "domain") %>% 
  mutate(domain_code = recode(name, 
                              "B-box-type zinc finger" = "Z",
                              "Zinc finger, RanBP2-type" = "Z",
                              "Zinc finger, ZZ-type" = "Z",
                              "Zinc finger, TAZ-type" = "Z",
                              "Zinc finger, TRAF-type" ="Z",
                              "Zinc finger, RING-type" = "Z",
                              "Zinc finger, SIAH-type" = "Z",
                              "U box domain" = "Z",
                              "BTB/POZ domain" = "B", 
                              "MATH/TRAF domain" = "T",
                              "Seven-in-absentia protein, TRAF-like domain" = "T",
                              "Peptidase C19, ubiquitin carboxyl-terminal hydrolase" = "P", 
                              "Ubiquitin carboxyl-terminal hydrolase 7, ICP0-binding domain" = "P",
                              "Ubiquitin carboxyl-terminal hydrolase, C-terminal" = "P",
                              "Ubiquitin specific protease domain" = "P",
                              .default = "")) %>%
  mutate(mid = (end+start)/2) %>%
  group_by(uniprot_accession) %>%
  arrange(mid, .by_group = TRUE) %>%
  mutate(protein_type2 = paste0(domain_code, collapse = "")) %>%
  mutate(protein_type2 = str_replace_all(protein_type2, "(P)\\1+", "\\1")) %>%
  ungroup() %>%
  dplyr::select(uniprot_accession, protein_type2) %>%
  distinct(uniprot_accession, .keep_all = TRUE)

# as.data.frame(table(as.factor(pdl_type2$name)))
pdl_withtype <- pdl %>%
  left_join(pdl_type) %>%
  left_join(pdl_type2)  %>%
  mutate(TEST = ifelse(protein_type == protein_type2,"","FIX ME")) %>%
  rowwise() %>% 
  mutate(protein_type_ = c_across(starts_with("protein_type"))[which.max(nchar(c_across(starts_with("protein_type"))))]) %>%
  mutate(ALL_have_T = str_match(protein_type_, "T")) %>% #check if all have T
  mutate(domain_have_T = str_match(protein_type2, "T")) %>%
  mutate(HSF_have_T = str_match(protein_type, "T"))


url_list_fasta<- tibble(ID= data,
                   URL = rep("https://www.uniprot.org/uniprot/REPLACE&format=fasta", length(data))) %>%
  mutate(URL = str_replace(URL, "REPLACE", ID))

res_f <- map(url_list_fasta$URL, GET)
table(unlist(map(res_f, c(status_code))))
# res_f <- discard(res_f, ~.x$status_code == 204)
saveRDS(res_f, "res_f.RDS")
res_f<- readRDS("res_f.RDS")


fasta <- map(res_f, "content") %>%
  map(~rawToChar(.)) %>%
  purrr::flatten() %>%
  as.character()

# fileconn <- file("May28.fasta")
# writeLines(fasta, fileconn)
# close(fileconn)


fasta2 <- map(res_f, "content") %>%
  map(~rawToChar(.)) %>%
  transpose() %>%
  as_tibble(.name_repair = "unique") %>%
  separate(`...1`, into = c("one", "two", "three"), sep="\\|") %>%
  separate(three, into = c("four","seq"), sep= "\n", extra = "merge") %>%
  separate(four, into = c("four", "species"), sep = "=", extra = "merge") %>%
  separate(species, into = c("species", "leftover"), sep = "OX=", extra = "merge") %>%
  mutate(seq = str_replace_all(seq, "\n", "")) %>%
  rename(uniprot_accession = two)

# pdl_TRAF <- pdl_withtype %>%
#   group_by(uniprot_accession) %>% 
#   # dplyr::filter(accession == "IPR018121") %>%
#   # nrow() #count number of SINA = 82
#   dplyr::filter(any(accession == "IPR018121")) %>%
#   dplyr::filter(accession == "IPR008974") %>%
#   nrow() # count number of HSF TRAF = 82 --> each SINA should be covered by HSF --> can exclude the domain


pdl_TRAF <- pdl_withtype %>%
  group_by(uniprot_accession) %>% 
  dplyr::filter(
    (accession == "IPR002083" & uniprot_accession != "A0A7I4D1F8" & uniprot_accession !="A0A2K1L2Z6") |  #TRAF domain for all but misannotated ones
      ((uniprot_accession == "A0A7I4D1F8"| uniprot_accession == "A0A2K1L2Z6") & accession == "IPR008974") | #add HSF for misannotated ones
      (any(accession == "IPR018121") & accession == "IPR008974") | #add HSF for SINA ones
      (is.na(domain_have_T) & accession == "IPR008974")) %>% #add HSF for ones missing TRAF domain annotation
  group_by(uniprot_accession) %>%
  mutate(id = 1:n()) %>%
  mutate(n = n())%>%
  ungroup() %>%
  mutate(domain_length = end-start) %>%
  mutate(T_n = str_count(protein_type_, "T")) %>%
  # filter(n != T_n) #966

##This code is for using mostly HSF instead
# pdl_TRAF <- pdl_withtype %>%
#   dplyr::filter(accession == "IPR008974" | (is.na(HSF_have_T) & accession == "IPR002083")) %>%
#   group_by(uniprot_accession) %>% 
#   mutate(id = 1:n()) %>%
#   mutate(n = n())%>%
#   ungroup() %>%
#   mutate(domain_length = end-start) %>%
#   mutate(T_n = str_count(protein_type_, "T")) %>%
#   filter(n != T_n)


pdl_fasta <- left_join(pdl_TRAF, fasta2, by = "uniprot_accession") %>%
  mutate(start = start-10)%>%
  mutate(end= end+10) %>%
  mutate(start = if_else(start <0, 0, start)) %>%
  mutate(end = if_else(end > as.numeric(protein_length), as.numeric(protein_length), end)) %>%
  # mutate(test = protein_length - end)
  mutate(subseq = str_sub(seq, start = start, end = end)) %>%
  mutate(species_short = word(species, 1, 2)) %>%
  unite(trace, id, n, sep= "/", remove = FALSE) %>%
  unite(trace_ua, uniprot_accession, trace, sep= "_", remove = FALSE) %>%
  unite(fasta_name, trace_ua, protein_type_, species_short, sep= " ", remove = FALSE)
  # mutate(fasta_name = str_replace(fasta_name, "_", "of")) %>%
  # separate(species, into =c("genus", "species", "subspeciesname"), sep = " ", remove = FALSE, extra = "merge")

# write_csv(pdl_fasta, "pdl_fasta.csv")

write.fasta(as.list(pdl_fasta$subseq), names= pdl_fasta$fasta_name, file.out = "June5subseq.fasta", as.string = TRUE)


#STILL HAVE TO GET RID OF DUPLICATES & SHORT SEQUENCES


#str(unique(pdl_TRAF$uniprot_accession)) #729

res_genes <- map(c('MGI_ID','SGD_ID','WORMBASE_ID','ENSEMBLGENOME_ID'),
                 ~POST(url = "https://www.uniprot.org/uploadlists/",
                      body = list(from = 'ACC+ID',
                        to = .x,
                        format = 'tab',
                        query = paste0(unique(pdl_TRAF$uniprot_accession),collapse = " ")
                        )
                      )
)

up_2_genes <- map_dfr(res_genes, ~content(.x, type = 'text/tab-separated-values')) %>% 
# up_2_genes[[3]]$To <- up_2_genes[[3]]$To %>% as.character()
# up_2_genes <- up_2_genes %>%
  # reduce(rbind) %>%
  rename(uniprot_accession = From, Gene_ID = To) %>%
  distinct(uniprot_accession, Gene_ID, .keep_all = TRUE)


up_2_genes2 <- content(
  POST(url = "https://www.uniprot.org/uploadlists/",
                      body = list(from = 'ACC+ID',
                        to = 'GENENAME',
                        format = 'tab',
                        query = (pdl_fasta %>% 
                                   dplyr::filter(species == "Marchantia polymorpha subsp. ruderalis " | species == "Papaver somniferum " | species == "Asparagus officinalis ") %>% 
                                   distinct(uniprot_accession) %>% 
                                   pull() %>% 
                                   paste0(collapse = " "))
                    ))
  , type = 'text/tab-separated-values') %>% 
  rename(uniprot_accession = From, Gene_ID = To) %>%
  distinct(uniprot_accession, Gene_ID, .keep_all = TRUE)

up_2_genes3 <- content(
  POST(url = "https://www.uniprot.org/uploadlists/",
                      body = list(from = 'ACC+ID',
                        to = 'GENENAME',
                        format = 'tab',
                        query = (left_join(pdl_fasta, rbind(up_2_genes, up_2_genes2), by = "uniprot_accession") %>% 
                                   filter(is.na(Gene_ID)) %>% 
                                   distinct(uniprot_accession) %>% 
                                   pull() %>% 
                                   paste0(collapse = " "))
                    ))
  , type = 'text/tab-separated-values') %>% 
  rename(uniprot_accession = From, Gene_ID = To) %>%
  distinct(uniprot_accession, Gene_ID, .keep_all = TRUE)



mappings <- rbind(up_2_genes, up_2_genes2, up_2_genes3)  %>%
  mutate(Gene_ID = str_replace(Gene_ID, "LOC_", "")) %>%
  # group_by(uniprot_accession) %>%
  # filter(n()>1) # some uniprot accession map to multiple gene IDs, but doesn't matter so much, since we care more about the other way around.
  group_by(Gene_ID) %>%
  filter(n()>1) %>%
  left_join(pdl_fasta) %>%
  arrange(Gene_ID) %>%
  ungroup() %>%
  distinct(Gene_ID)

pdl_fasta_gID <- left_join(pdl_fasta, rbind(up_2_genes, up_2_genes2, up_2_genes3), by = "uniprot_accession") %>%
  mutate(Gene_ID = str_replace(Gene_ID, "LOC_", ""))
  group_by(species, subseq, Gene_ID) %>%
  filter(n()>1)


no_ID <- anti_join(pdl_TRAF, up_2_genes, by = "uniprot_accession") %>%
  distinct(uniprot_accession, .keep_all = TRUE)

# 
# 
# 
# url_list_gene<- tibble(ID= data,
#                    URL = rep("https://www.uniprot.org/uniprot/REPLACE&format=txt", length(data))) %>%
#   mutate(URL = str_replace(URL, "REPLACE", ID))
# 
# res_gene <- map(url_list_gene$URL, GET)
# table(unlist(map(res_gene, c(status_code))))
# 
# saveRDS(res_gene, "res_gene.RDS")
# res_gene<- readRDS("res_gene.RDS")
# 
# genes <- map(res_gene, "content") %>%
#   map(~rawToChar(.))
#   # purrr::flatten() %>%
#   # as.character()
# 
# 
# gene_annot <- map(genes, ~as_tibble(read_lines(.x))) %>%
#   map(~filter(.x, str_detect(.x$value, "DR   MGI|DR   EnsemblMetazoa|DR   WormBase|DR   EnsemblFungi|DR   EnsemblPlants")) %>% transpose())










  
```

##look up duplicates
```{r}


ensemblp <- useEnsemblGenomes(biomart = "plants_mart")
ensemblp_datasets <- as_tibble(listDatasets(ensemblp)) %>%
  separate(description, into = c("genus", "species", "speciesdetails"), sep = " ", extra = "merge")

species_to_search <- distinct(pdl_fasta, genus, species, subspeciesname) %>% dplyr::select(genus,species,subspeciesname) %>%
  left_join(ensemblp_datasets) %>%
  drop_na(dataset) %>%
  filter(!str_detect(dataset, "oindica_eg_gene")) %>%
  pull(dataset)

pdl_fasta_ <- left_join(pdl_fasta, species_to_search) %>%
  drop_na(dataset) %>%
  group_by(dataset)

pdl_fasta_split <- pdl_fasta_ %>%
  group_split(.keep=TRUE)

search_biomart_filters <- function(dataset){
  as_tibble(
    rbind(
    searchAttributes(useEnsemblGenomes(biomart = "plants_mart", dataset = dataset), pattern ="uniprot"),
    searchAttributes(useEnsemblGenomes(biomart = "plants_mart", dataset = dataset), pattern ="swiss")
  ))
}

searchAttributes(useEnsemblGenomes(biomart = "plants_mart", dataset = "athaliana_eg_gene"), pattern ="uniprot")
biomart_filters <- map(species_to_search$dataset, search_biomart_filters)





Filters <- "uniprotsptrembl"
Attributes <- c("ensembl_gene_id","ensembl_transcript_id", "uniprotsptrembl")

Search_biomart <- function(search_terms, biomart = "plants_mart", attributes = Attributes, filters = Filters){
  mart_ID <- useEnsemblGenomes(biomart = biomart, dataset = search_terms$dataset[[1]])
  as_tibble(
    getBM(attributes = attributes,
      filters = filters,
      values = search_terms$uniprot_accession, 
      mart = mart_ID)
  )
}
Ensembl_annot <- map(pdl_fasta_split, Search_biomart)
pdl_fasta_split <- pdl_fasta_split[-4]
Ensembl_annot2 <- map(pdl_fasta_split, Search_biomart, filters = "uniprotswissprot", attributes = c("ensembl_gene_id","ensembl_transcript_id", "uniprotswissprot"))
Ensembl_annot_full <- c(Ensembl_annot, Ensembl_annot2) %>%
  discard(function(x) nrow(x) == 0) %>%
  reduce(full_join, by = "ensembl_gene_id")


#TIR_list <- map(dataset_id, Search_biomart) #if looking through multiple species
#str(TIR_list)
#TIR_list <- TIR_list$athaliana_eg_gene 

TIR_list <- Search_biomart(dataset_id)
```
