---
title: "TRAF_for_class"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("biomaRt")
install.packages("tidyverse")
install.packages(c("httr","jsonlite"))
install.packages("seqinr")

library(tidyverse)
library(biomaRt)
library(httr)
library(jsonlite)
library(seqinr)

#Get genes
IDs <- scan("diverse_proteomes.txt", character())
url_list <- tibble(ID = IDs,
                   URL = rep("https://www.ebi.ac.uk/interpro/api/protein/UniProt/entry/InterPro/IPR008974/proteome/uniprot/REPLACE/?page_size=300",length(IDs)),
                   URL2 = rep("https://www.ebi.ac.uk/interpro/api/protein/UniProt/entry/InterPro/IPR002083/proteome/uniprot/REPLACE/?page_size=300",length(IDs))
                   ) %>%
  mutate(URL = str_replace(URL, "REPLACE", ID)) %>%
  mutate(URL2 = str_replace(URL2, "REPLACE", ID))


res_ <- map(c(url_list$URL,url_list$URL2), GET)

saveRDS(res_, "res_.RDS")
res_<- readRDS("res_.RDS")

table(unlist(map(res_, c(status_code))))
# res_ <- discard(res_, ~.x$status_code == 204) #remove lists that don't have any data if any
data <- map(res_, "content") %>%
  map(~fromJSON(rawToChar(.))) %>%
  map(c("results","metadata","accession")) %>%
  unlist() %>%
  unique() # generates list of uniprot accessions containing interpro accessions - NOTE THAT SOME ISOFORMS MAY NOT BE INCLUDED

# IDs2 <- tibble(read_delim("tax_IDs_remainder.txt", col_names = FALSE, delim= "\n",col_types = "c"))
# url_list_ <- tibble(ID = unlist(IDs2),
#                    URL = rep("https://www.ebi.ac.uk/interpro/api/protein/UniProt/entry/InterPro/IPR008974/taxonomy/uniprot/REPLACE/?page_size=300",length(IDs2))) %>% 
#     mutate(URL = str_replace(URL, "REPLACE", ID))
# 
# res__ <- map(url_list_$URL, GET)
# res__ <- discard(res__, ~.x$status_code == 204)
# data2 <- map(res__, "content") %>%
#   map(~fromJSON(rawToChar(.))) %>%
#   map(c("results","metadata","accession")) %>%
#   purrr::flatten() %>%
#   as.character()%>%
#   as_tibble() %>%
#   distinct()
# 
# data_all <- rbind(data,data2)


url_list3 <- tibble(Accession = data, URL = rep("https://www.ebi.ac.uk/interpro/api/entry/InterPro/protein/UniProt/REPLACE?page_size=350",length(data))) %>%
  mutate(URL = str_replace(URL, "REPLACE", Accession)) 

res_3 <- map(url_list3$URL, GET) #get interpro data for each accession
table(unlist(map(res_3, c(status_code))))
# res_3 <- discard(res_3, ~.x$status_code == 204)

saveRDS(res_3, "res_3.RDS")
res_3<- readRDS("res_3.RDS")


meta <- map(res_3, "content") %>%
  map(~fromJSON(rawToChar(.))) %>%
  map("results")
  

itp_meta <- meta %>%
  map(c("metadata")) %>%
  map(~dplyr::select(.x, "accession","name","source_database","type")) %>%
  map(transpose)

itp_prot <- meta %>%
  map(c("proteins")) %>%
  map(~map(.x, ~dplyr::rename(.x, "uniprot_accession" = "accession"))) %>%
  map(~map(.x, ~dplyr::rename(.x, "uniprot_source_database" = "source_database"))) %>%
  map(~map(.x, purrr::flatten)) %>%
  map(~map(.x, purrr::flatten))

itp_ <-map2(itp_meta, itp_prot, ~map2(.x,.y, c))

itp_names <- itp_ %>%
  map(~map(., names)) %>%
  map(unlist) %>%
  purrr::reduce(union)

itp_ <- itp_ %>%
  map(transpose, .names = itp_names) %>%
  # map(~map_if(.x, is.data.frame, list)) %>%
  map(~as_tibble(.x, .name_repair = "unique")) %>%
  map(~unnest(.x, c("fragments", "model", "score"))) %>%
  map(~unnest(.x, fragments)) %>%
  map(transpose)

itp_names2 <- itp_ %>%
  map(~map(., names)) %>%
  map(unlist) %>%
  purrr::reduce(union) # to make sure all columns are kept

itp__ <- itp_ %>%
  map(transpose, .names = itp_names2) %>%
  map(~as_tibble(.x, .name_repair = "unique"))

pdl <- bind_rows(itp__, .id="Order") %>%
  unnest(cols = colnames(.)) %>%
  mutate(uniprot_accession = toupper(uniprot_accession))

# pdl_map <- pdl %>%
#   group_by(Order, uniprot_accession) %>%
#   summarise() %>%
#   drop_na(uniprot_accession)

pdl_type <- pdl %>%
  filter(type == "homologous_superfamily") %>%
  mutate(domain_code = recode(accession, "IPR008974" = "T", "IPR011333" = "B", "IPR013083" = "Z", "IPR038765" = "P", .default = "")) %>%
  mutate(mid = (end+start)/2) %>%
  group_by(uniprot_accession) %>%
  arrange(mid, .by_group = TRUE) %>%
  mutate(protein_type = paste0(domain_code, collapse = "")) %>%
  ungroup() %>%
  dplyr::select(uniprot_accession, protein_type) %>%
  distinct(uniprot_accession, .keep_all = TRUE)

pdl_type2 <- pdl %>%
  filter(type == "domain") %>% 
  mutate(domain_code = recode(name, 
                              "B-box-type zinc finger" = "Z",
                              "Zinc finger, RanBP2-type" = "Z",
                              "Zinc finger, ZZ-type" = "Z",
                              "Zinc finger, TAZ-type" = "Z",
                              "Zinc finger, TRAF-type" ="Z",
                              "Zinc finger, RING-type" = "Z",
                              "Zinc finger, SIAH-type" = "Z",
                              "U box domain" = "Z",
                              "BTB/POZ domain" = "B", 
                              "MATH/TRAF domain" = "T",
                              "Seven-in-absentia protein, TRAF-like domain" = "T",
                              "Peptidase C19, ubiquitin carboxyl-terminal hydrolase" = "P", 
                              "Ubiquitin carboxyl-terminal hydrolase 7, ICP0-binding domain" = "P",
                              "Ubiquitin carboxyl-terminal hydrolase, C-terminal" = "P",
                              "Ubiquitin specific protease domain" = "P",
                              .default = "")) %>%
  mutate(mid = (end+start)/2) %>%
  group_by(uniprot_accession) %>%
  arrange(mid, .by_group = TRUE) %>%
  mutate(protein_type2 = paste0(domain_code, collapse = "")) %>%
  mutate(protein_type2 = str_replace_all(protein_type2, "(P)\\1+", "\\1")) %>%
  ungroup() %>%
  dplyr::select(uniprot_accession, protein_type2) %>%
  distinct(uniprot_accession, .keep_all = TRUE)

# as.data.frame(table(as.factor(pdl_type2$name)))
pdl_withtype <- pdl %>%
  left_join(pdl_type) %>%
  left_join(pdl_type2)  %>%
  mutate(TEST = ifelse(protein_type == protein_type2,"","FIX ME")) %>%
  rowwise() %>% 
  mutate(protein_type_ = c_across(starts_with("protein_type"))[which.max(nchar(c_across(starts_with("protein_type"))))]) %>%
  mutate(ALL_have_T = str_match(protein_type_, "T")) %>% #check if all have T
  mutate(domain_have_T = str_match(protein_type2, "T")) %>%
  mutate(HSF_have_T = str_match(protein_type, "T"))


url_list_fasta<- tibble(ID= data,
                   URL = rep("https://www.uniprot.org/uniprot/REPLACE&format=fasta", length(data))) %>%
  mutate(URL = str_replace(URL, "REPLACE", ID))

res_f <- map(url_list_fasta$URL, GET)
table(unlist(map(res_f, c(status_code))))
# res_f <- discard(res_f, ~.x$status_code == 204)
saveRDS(res_f, "res_f.RDS")
res_f<- readRDS("res_f.RDS")


fasta <- map(res_f, "content") %>%
  map(~rawToChar(.)) %>%
  purrr::flatten() %>%
  as.character()

# fileconn <- file("May28.fasta")
# writeLines(fasta, fileconn)
# close(fileconn)


fasta2 <- map(res_f, "content") %>%
  map(~rawToChar(.)) %>%
  transpose() %>%
  as_tibble(.name_repair = "unique") %>%
  separate(`...1`, into = c("one", "two", "three"), sep="\\|") %>%
  separate(three, into = c("four","seq"), sep= "\n", extra = "merge") %>%
  separate(four, into = c("four", "species"), sep = "=", extra = "merge") %>%
  separate(species, into = c("species", "leftover"), sep = "OX=", extra = "merge") %>%
  mutate(seq = str_replace_all(seq, "\n", "")) %>%
  rename(uniprot_accession = two)

# pdl_TRAF <- pdl_withtype %>%
#   group_by(uniprot_accession) %>% 
#   # dplyr::filter(accession == "IPR018121") %>%
#   # nrow() #count number of SINA = 82
#   dplyr::filter(any(accession == "IPR018121")) %>%
#   dplyr::filter(accession == "IPR008974") %>%
#   nrow() # count number of HSF TRAF = 82 --> each SINA should be covered by HSF --> can exclude the domain


pdl_TRAF <- pdl_withtype %>%
  group_by(uniprot_accession) %>% 
  dplyr::filter(
    (accession == "IPR002083" & uniprot_accession != "A0A7I4D1F8" & uniprot_accession !="A0A2K1L2Z6") |  #TRAF domain for all but misannotated ones
      ((uniprot_accession == "A0A7I4D1F8"| uniprot_accession == "A0A2K1L2Z6") & accession == "IPR008974") | #add HSF for misannotated ones
      (any(accession == "IPR018121") & accession == "IPR008974") | #add HSF for SINA ones
      (is.na(domain_have_T) & accession == "IPR008974")) %>% #add HSF for ones missing TRAF domain annotation
  group_by(uniprot_accession) %>%
  mutate(id = 1:n()) %>%
  mutate(n = n())%>%
  ungroup() %>%
  mutate(domain_length = end-start) %>%
  mutate(T_n = str_count(protein_type_, "T")) #%>%
  # filter(n != T_n) #966

##This code is for using mostly HSF instead
# pdl_TRAF <- pdl_withtype %>%
#   dplyr::filter(accession == "IPR008974" | (is.na(HSF_have_T) & accession == "IPR002083")) %>%
#   group_by(uniprot_accession) %>% 
#   mutate(id = 1:n()) %>%
#   mutate(n = n())%>%
#   ungroup() %>%
#   mutate(domain_length = end-start) %>%
#   mutate(T_n = str_count(protein_type_, "T")) %>%
#   filter(n != T_n)


pdl_fasta <- left_join(pdl_TRAF, fasta2, by = "uniprot_accession") %>%
  mutate(subseq_4_compare = str_sub(seq, start = start, end = end)) %>%
  mutate(start = start-10)%>%
  mutate(end= end+10) %>%
  mutate(start = if_else(start <0, 0, start)) %>%
  mutate(end = if_else(end > as.numeric(protein_length), as.numeric(protein_length), end)) %>%
  # mutate(test = protein_length - end)
  mutate(subseq = str_sub(seq, start = start, end = end)) %>%
  mutate(species_short = word(species, 1, 2)) %>%
  unite(trace, id, n, sep= "/", remove = FALSE) %>%
  unite(trace_ua, uniprot_accession, trace, sep= "_", remove = FALSE) %>%
  unite(fasta_name, trace_ua, protein_type_, species_short, sep= " ", remove = FALSE) %>%
  mutate(simp_type = if_else(str_detect(protein_type_, "B"),"Z",
                             if_else(str_detect(protein_type_, "Z"),"Z",
                                     if_else(str_detect(protein_type_, "P"), "P",
                                             if_else(str_detect(protein_type_, "TTTT"), "4T", "T"))))) %>%
  mutate(tip = row_number()-1) %>%
  mutate(t = "t")%>%
  unite("tip", "t","tip", sep = "") %>%
  relocate(tip)
  # mutate(fasta_name = str_replace(fasta_name, "_", "of")) %>%
  # separate(species, into =c("genus", "species", "subspeciesname"), sep = " ", remove = FALSE, extra = "merge")

# write_csv(pdl_fasta, "pdl_fasta.csv")

# write_csv(pdl_fasta, "pdl_fasta.csv")
# write.fasta(as.list(pdl_fasta$subseq), names= pdl_fasta$fasta_name, file.out = "June5subseq.fasta", as.string = TRUE)


#STILL HAVE TO GET RID OF DUPLICATES & SHORT SEQUENCES


#str(unique(pdl_TRAF$uniprot_accession)) #729

res_genes <- map(c('MGI_ID','SGD_ID','WORMBASE_ID','ENSEMBLGENOME_ID'),
                 ~POST(url = "https://www.uniprot.org/uploadlists/",
                      body = list(from = 'ACC+ID',
                        to = .x,
                        format = 'tab',
                        query = paste0(unique(pdl_TRAF$uniprot_accession),collapse = " ")
                        )
                      )
)

up_2_genes <- map_dfr(res_genes, ~content(.x, type = 'text/tab-separated-values')) %>% 
# up_2_genes[[3]]$To <- up_2_genes[[3]]$To %>% as.character()
# up_2_genes <- up_2_genes %>%
  # reduce(rbind) %>%
  rename(uniprot_accession = From, Gene_ID = To) %>%
  distinct(uniprot_accession, Gene_ID, .keep_all = TRUE)


up_2_genes2 <- content(
  POST(url = "https://www.uniprot.org/uploadlists/",
                      body = list(from = 'ACC+ID',
                        to = 'GENENAME',
                        format = 'tab',
                        query = (pdl_fasta %>% 
                                   dplyr::filter(species == "Marchantia polymorpha subsp. ruderalis " | species == "Papaver somniferum " | species == "Asparagus officinalis ") %>% 
                                   distinct(uniprot_accession) %>% 
                                   pull() %>% 
                                   paste0(collapse = " "))
                    ))
  , type = 'text/tab-separated-values') %>% 
  rename(uniprot_accession = From, Gene_ID = To) %>%
  distinct(uniprot_accession, Gene_ID, .keep_all = TRUE)

up_2_genes3 <- content(
  POST(url = "https://www.uniprot.org/uploadlists/",
                      body = list(from = 'ACC+ID',
                        to = 'GENENAME',
                        format = 'tab',
                        query = (left_join(pdl_fasta, rbind(up_2_genes, up_2_genes2), by = "uniprot_accession") %>% 
                                   filter(is.na(Gene_ID)) %>% 
                                   distinct(uniprot_accession) %>% 
                                   pull() %>% 
                                   paste0(collapse = " "))
                    ))
  , type = 'text/tab-separated-values') %>% 
  rename(uniprot_accession = From, Gene_ID = To) %>%
  distinct(uniprot_accession, Gene_ID, .keep_all = TRUE)


gID_mappings <- rbind(up_2_genes, up_2_genes2, up_2_genes3)
saveRDS(gID_mappings, "gID_mappings.RDS")
gID_mappings <- readRDS("gID_mappings.RDS")


#get gene IDS for all sequences
pdl_fasta_gID <- gID_mappings  %>%
  mutate(Gene_ID = str_replace(Gene_ID, "LOC_", "")) %>%
  mutate(Gene_ID = str_replace(Gene_ID, "v3$", "")) %>%
  # group_by(uniprot_accession) %>%
  # filter(n()>1) # some uniprot accession map to multiple gene IDs, but doesn't matter so much, since we care more about the other way around.
  group_by(Gene_ID) %>%
  mutate(isoform_problem = if_else(n()>1,"Yes", "No")) %>%
  group_by(uniprot_accession) %>%
  mutate(duplicated_UA = if_else(n()>1,"Yes", "No"))%>%
  right_join(pdl_fasta) %>%
  relocate(tip) %>%
  mutate(uniprot_accession_l = nchar(uniprot_accession)) # note this includes 8 situations where one uniprot matches to multiple gene IDs

# saveRDS(pdl_fasta_gID, "pdl_fasta_gID.RDS")
# pdl_fasta_gID<- readRDS("pdl_fasta_gID.RDS")


pdl_fasta_gID <- pdl_fasta_gID %>%
  group_by(Gene_ID) %>%
  arrange(nchar(uniprot_accession), .by_group = TRUE) %>% # to keep shorter accession
  ungroup() %>%
  group_by(Gene_ID, uniprot_accession) %>%
  # arrange(trace_ua, .by_group = TRUE) %>%
  mutate(subseq_c = paste0(subseq_4_compare, collapse = "_")) %>% # collapse TRAF domains from each uniprot_accession into one
  ungroup() %>%
  group_by(Gene_ID) %>%
  arrange(nchar(uniprot_accession), .by_group = TRUE) %>% #to keep the identical sequence  with the shorter accession name
  # group_by(Gene_ID) %>%
  distinct(trace, subseq_c, .keep_all = TRUE)%>%
  ungroup() 
  

#delete isoforms that have less than all domains  
UA_2_del <- pdl_fasta_gID %>%  
  group_by(Gene_ID) %>%
  filter(n()>1) %>%
  distinct(n, .keep_all = TRUE) %>%
  group_by(Gene_ID) %>%
  filter(n()>1) %>%
  group_by(Gene_ID) %>%
  arrange(desc(n), .by_group = TRUE) %>%
  ungroup() %>%
  distinct(Gene_ID, .keep_all = TRUE) %>%
  dplyr::select(uniprot_accession, Gene_ID, n)


pdl_fasta_gID_ <- anti_join(pdl_fasta_gID, UA_2_del, by = "uniprot_accession") %>%
  # distinct(uniprot_accession, .keep_all = TRUE) %>%
  # group_by(seq,subseq_c) %>%
  group_by(Gene_ID, trace) %>%
  # filter(n()>1) %>%
  # distinct(Gene_ID)
  mutate(isoform_to_man_remove = if_else(n()>1,"Yes", "No")) %>%
  ungroup() %>%
  mutate(domain_size = if_else(domain_length <90, "SHORT", "OK"))

    # group_by(uniprot_accession) %>%
  # filter(n()>1) # some uniprot accession map to multiple gene IDs, but doesn't matter so much, since we care more about the other way around.


#delete isoforms that have short domain
UA_2_del2 <- pdl_fasta_gID_%>%  
  group_by(Gene_ID, trace) %>%
  filter(n()>1) %>%
  ungroup() %>% group_by(uniprot_accession) %>%
  mutate(short = if_else(any(domain_size == "SHORT"), "SHORT", "NORMAL")) %>%
  ungroup() %>% group_by(Gene_ID) %>%
  # filter(!any(short == "NORMAL")) #check if any only have "short" sequences - only one - remove A0A2K1XQX7, keep A0A2K1XQY0
  filter(short == "SHORT") %>%
  filter(uniprot_accession != "A0A2K1XQY0") %>%
  distinct(uniprot_accession)

pdl_fasta_gID_ <- anti_join(pdl_fasta_gID_, UA_2_del2, by = "uniprot_accession") #%>%  
  # group_by(Gene_ID, trace) %>%
  # filter(n()>1) %>%
  # ungroup() %>% group_by(Gene_ID) %>%
  # relocate(domain_length, start, end, protein_length)

#this section for manually check which isoforms should be removed

tree_annot <- pdl_fasta_gID_ %>%
  ungroup() %>%
  arrange(nchar(Gene_ID))%>%
  # group_by(tip)%>% 
  # filter(n()>1)
  distinct(tip, .keep_all = TRUE) %>%  #it's necessary to remove duplicated tips at this point since haven't removed some dups yet
  unite(UP_ID,  uniprot_accession, Gene_ID, trace, protein_type_, species_short, remove = FALSE)

tax_TT <- read_tsv("RAxML-2021.06.05-Run.1 isoforms included/taxonNamesTranslationTable.txt", col_names = c("tip.label", "fasta_name"))
tree_annot <- left_join(tree_annot,tax_TT, by = "fasta_name") %>%
  relocate(tip.label)
````



```{r}
BiocManager::install("ggtree")
BiocManager::install("treeio")
install.packages("svglite")
install.packages("geiger")
install.packages("phangorn")
install.packages("phytools")
install.packages("Polychrome")
install.packages("ggnewscale")
library(ggtree)
library(treeio)
library(ape)
library(svglite)
library(geiger)
library(phangorn)
library(phytools)
library(Polychrome)
library(ggnewscale)

raxml <- read.newick("RAxML-2021.06.05-Run.1 isoforms included/RAxML_bestTree.file.out")
raxml_annot <- ggtree(raxml, layout = "circular") %<+% tree_annot +
  geom_tiplab(aes(label= UP_ID, color = simp_type), size = 2) +
  geom_tippoint(aes(fill = species), shape = 21, size = 2) 

raxml_annot_ <- gheatmap(raxml_annot, tree_annot %>% column_to_rownames(var = "tip.label") %>%dplyr::select(species), offset= 1, width = 0.03, colnames = FALSE)
ggsave("raxml_annotated_old_fixed.pdf", plot = raxml_annot_,
       device = "pdf", 
       width = 1000, 
       height =1000, 
       units = "mm",
       dpi=300,
       limitsize = FALSE)

#delete isoforms - manual annotation
UA_2_del3 <- tibble(uniprot_accession = strsplit("A8MQL1 F4I476 A0A1P8ARC9 A0A1P8AX66 O48777 F4JDF6 F4J5T4 A0A1P8B3G0 F4JKT0 A8MR02 F4KG56 A0A1P8BC00 A0A0A6YWE7 A0A0A6YXL9 Q6A4J8 Q717B2 Q6YCH1 Q6Z6Z3 Q0ZCE4 U7DYS1 A0A2K1X3E1 A0A3N7G431 A0A3N7G464  A0A3N7G7H1 A0A3N7HVW6 A0A7I4BK46 A0A7I4F8L4 A0A7I4EDU9 G3MU79 Q17782 P34324 A0A3N7FDW1 A0A2K1ZES4 A0A3N7FP35 B9I5C4", " ")[[1]], by = "uniprot_accession")


pdl_fasta_gID_ <- anti_join(pdl_fasta_gID_, UA_2_del3, by = "uniprot_accession")

#delete genes that have identical sequences (uniprot IDs mapping to multiple genes)
UA_2_del4 <- pdl_fasta_gID_ %>%  
  group_by(uniprot_accession, trace) %>%
  filter(n()>1) %>%
  arrange(nchar(Gene_ID)) %>%
  ungroup() %>% group_by(uniprot_accession) %>%
  slice(1)

pdl_fasta_gID_ <- anti_join(pdl_fasta_gID_, UA_2_del4, by = "uniprot_accession") #%>%
  # group_by(subseq, seq) %>%
  # filter(n()>1)  #13 subseq are duplicated, but they come from different genes that have a different full-length sequence - so probably a recent duplication?



tree_annot <- pdl_fasta_gID_ %>%
  ungroup() %>% #group_by(Gene_ID, trace, tip) %>% filter(n()>1) #check if any dup tips
  arrange(nchar(Gene_ID))%>%
  # distinct(tip, .keep_all = TRUE) %>% #only if necessary
  unite(UP_ID,  uniprot_accession, Gene_ID, trace, protein_type_, species_short, remove = FALSE)

write_csv(tree_annot, "pdl_fasta.csv")
write.fasta(as.list(tree_annot$subseq), names= tree_annot$UP_ID, file.out = "June9subseq_noiso.fasta", as.string = TRUE)
tree_annot <- read.csv("pdl_fasta.csv")



raxml <- read.newick("RAxML-2021.06.08-Run.1/RAxML_bestTree.result")


raxml_annot <- ggtree(raxml, layout = "circular") %<+% tree_annot +
  geom_tiplab(aes(label= UP_ID, color = species_short), size = 2) +
  geom_tippoint(aes(fill = simp_type), shape = 21, size = 2) 

# raxml_annot_ <- gheatmap(raxml_annot, tree_annot %>% column_to_rownames(var = "tip") %>%dplyr::select(simp_type), offset= 1, width = 0.03, colnames = FALSE)


ggsave("raxml_annotated2.pdf", plot = raxml_annot,
       device = "pdf", 
       width = 1000, 
       height =1000, 
       units = "mm",
       dpi=300,
       limitsize = FALSE)

#Look at IQ.TREE

#write iqtree without tip labels for bootstrap value matching thru RAxML
tree_annot <- read.csv("pdl_fasta.csv")
iqtree <- read.newick("IQ-tree_ML_treesearch2_2 best_tree/output.treefile")
tree_annot <- tree_annot %>% 
  mutate(a = row_number()-1) %>%
  mutate(b = "t") %>%
  unite("tip_aln", "b","a", sep = "") %>% 
  relocate(tip_aln) %>%
  rename(tip.label = UP_ID) %>%
  mutate(tip.label = str_replace(tip.label, " ", "_")) %>%
  mutate(tip.label = str_replace(tip.label, "\\:", "_"))%>%
  mutate(tip_a = tip_aln) %>%
  relocate(tip_aln, tip_a)


iqtree_mapping <- left_join(tibble(tip.label = iqtree$tip.label), (tree_annot %>% dplyr::select(tip.label, tip_aln)))


iqtree$tip.label <- iqtree_mapping$tip_aln
write.tree(iqtree, file = "IQ-tree_ML_treesearch2_2 best_tree/combined with boostrap/iqtree_notaxnames.besttree", digits = 15)
#now go do the mapping

#now read in the mapped file
iqtree_bs <- read.newick("IQ-tree_ML_treesearch2_2 best_tree/combined with boostrap/RAxML_bipartitions.iqtree_besttree_with_raxml_boostrap.phy")

tree_annot <- tree_annot %>%
  mutate(group = if_else(str_detect(species, "elegans|musculus|cerevisiae"), "Opisthokonta", "Archaeplastida")) %>%
  unite(Text_color_label, species_short, protein_type_, trace_ua, sep = " ", remove = FALSE)




iqtree_annot <- ggtree(iqtree_bs, layout="circular") %<+% tree_annot +
  geom_tiplab(aes(label= Text_color_label, color = group), size = 2) +
  geom_tippoint(aes(fill = domain_size), shape = 21, size = 2) +
  # geom_text(aes(label=node)) +
  geom_hilight(node = 790, alpha = 0.15, fill = "aquamarine", extendto = 4) +
    geom_hilight(node = 858, alpha = 0.15, fill = "limegreen", extendto = 4) +
      geom_hilight(node = 1003, alpha = 0.15, fill = "blue", extendto = 4) +
      geom_hilight(node = 1033, alpha = 0.15, fill = "chartreuse3", extendto = 4) +
      geom_hilight(node = 952, alpha = 0.15, fill = "slateblue4", extendto = 4) +
      geom_hilight(node = 1499, alpha = 0.15, fill = "darkgoldenrod", extendto = 4) +
      geom_hilight(node = 1049, alpha = 0.15, fill = "firebrick", extendto = 4) +
      geom_hilight(node = 1106, alpha = 0.15, fill = "deeppink4", extendto = 4) +

  
  
  geom_text2(aes(label=label, subset = !is.na(as.numeric(label)) & as.numeric(label) > 50),  size = 2) #hjust=2, vjust =1,

  
# iqtree_annot <- groupClade(iqtree_bs, c(790, 858, 1003, 1033, 952, 1499, 1049, 1106))
#   iqtree_annot <- ggtree(iqtree_bs, aes(color = group), layout =)
# 
# group 790 - TTTT
#   group 858 - 2T
#   group 1003 - ranun specific expansion
#   group 1033 - ranun tp clade
#   group 952 - TP group
#   group 1499 ??
#     group 1049 - 2 T group
#   group 1106 - BTB
#   
  
# raxml_annot_ <- gheatmap(raxml_annot, tree_annot %>% column_to_rownames(var = "tip.label") %>%dplyr::select(species), offset= 1, width = 0.03, colnames = FALSE)
# iqtree_annot

ggsave("IQ-TREE_best_tree_annot_foranal.pdf", plot = iqtree_annot,
       device = "pdf", 
       width = 1000, 
       height =1000, 
       units = "mm",
       dpi=300,
       limitsize = FALSE)


iqtree_bs_4 <- tree_subset(iqtree_bs, node = 790, levels_back = 1)

iqtree_annot4_ <- ggtree(iqtree_bs_4) %<+% tree_annot +
  geom_tiplab(aes(label= trace, color = trace), size = 2) +
  geom_tippoint(aes(fill = species), shape = 21, size = 2) +
  geom_text2(aes(label=label, subset = !is.na(as.numeric(label)) & as.numeric(label) > 50), hjust=2, vjust =2, size = 2)


ggsave("IQ-TREE_best_tree_annot4T_only.pdf", plot = iqtree_annot4_,
       device = "pdf", 
       width = 500, 
       height =1000, 
       units = "mm",
       dpi=300,
       limitsize = FALSE)










#now try with the RAxML_tree
raxml_bs <- read.newick("RAxML-2021.06.10-run.9GAMMA JTT F/combined with boostrap/RAxML_bipartitions.mapped")

raxml_bs_annot <- ggtree(raxml_bs) %<+% tree_annot +
  geom_tiplab(aes(label= tip.label, color = simp_type), size = 2) +
  geom_tippoint(aes(fill = species), shape = 21, size = 2) +
  # geom_text(aes(label=node))
  geom_text2(aes(label=label, subset = !is.na(as.numeric(label)) & as.numeric(label) > 50), hjust=2, vjust =1, size = 2)

# raxml_annot_ <- gheatmap(raxml_annot, tree_annot %>% column_to_rownames(var = "tip.label") %>%dplyr::select(species), offset= 1, width = 0.03, colnames = FALSE)
ggsave("RAxML_BS_best_tree_annot.pdf", plot = iqtree_annot,
       device = "pdf", 
       width = 1000, 
       height =1000, 
       units = "mm",
       dpi=300,
       limitsize = FALSE)








pdl_fasta_genelist <- tree_annot%>%
  group_by(Gene_ID)%>%
  distinct(Gene_ID, .keep_all = TRUE)
  

write_csv(pdl_fasta_genelist, "pdl_fasta_genegrouped.csv")












####Generate 4T single gene tree

#read phylip mapped file (not nexus)
iqtree_4T_ML <- read.tree(file = "addtax for 4T/IQ_TREE-2021.06.18-Run.4 THIS ONE/THIS ONE/RAxML_bipartitions.mapped_gfjhgfj")

species_factor <- read_delim("addtax for 4T/IQ_TREE-2021.06.18-Run.4 THIS ONE/THIS ONE/species_order_for_levels.txt", delim = "\t", col_names = "species_level")
species_factor <- as_factor(c(species_factor$species_level, "LOST"))

color_palette <- c(rainbow(32, start = 0.25, end = 0.1), "black")
# swatch(c(rainbow(32, start = 0.25, end = 0.1), "black"))
names(color_palette) <- species_factor

tipnames_4T <- read_delim("addtax for 4T/IQ_TREE-2021.06.18-Run.4 THIS ONE/THIS ONE/Taxnames_forR.txt", delim = "\t", col_names = c("Acc","taxon_no","notung_ID","species", "uniprot_accession")) %>%
  group_by(species) %>%
  mutate(code_name = paste0(species,"_QMATH_",letters[1:n()])) %>%
  mutate(species = factor(species, levels = levels(species_factor)))

Gene_4T <- ggtree(iqtree_4T_ML) %<+% tipnames_4T +
  # geom_tiplab() +
  # geom_text(aes(label=node)) +
  geom_rootedge(rootedge = 0.25) +
  geom_text2(aes(label=label, subset = !is.na(as.numeric(label)), color = !is.na(as.numeric(label)) & as.numeric(label) <75), hjust=1, vjust =-0.5, size = 2) +  scale_color_manual(values=c("black", "red"), guide = FALSE) +
  new_scale_color() +
  geom_tiplab(aes(label= code_name, color = species), size = 2) +
  scale_color_manual(values=color_palette)+ #unname(glasbey.colors(25)[2:25]) 
  guides(color=guide_legend(ncol=1, )) +
  theme(legend.title = element_text(size=5),
        legend.text = element_text(size=5),
        legend.position = "left",
        plot.margin = unit(c(0.5,4,3,0), "cm"))+
  coord_cartesian(clip = 'off')
        

Gene_4T


# swatch(rainbow(24, start = 0.25, end = 0.1))

ggsave("4T_gene_tree.pdf", plot = Gene_4T,
       device = "pdf", 
       width = 215.9, 
       height =279.4, 
       units = "mm",
       dpi=300)


####Generate Notung tree
iqtree_4T_notung <- read.nhx(file = "addtax for 4T/IQ_TREE-2021.06.18-Run.4 THIS ONE/THIS ONE/output.boottrees_for_notung_result.phy.NHX")
tipnames_4T_notung2 <- read_delim("addtax for 4T/IQ_TREE-2021.06.18-Run.4 THIS ONE/THIS ONE/Taxnames_forR2.txt", delim = "\t", col_names = TRUE)

tipnames_4T_notung <- rbind(tipnames_4T,tipnames_4T_notung2)%>%
  relocate(notung_ID) %>%
    mutate(species = factor(species, levels = c(levels(species_factor), "LOST")))

iqtree_4T_notung@data$D <- str_replace(iqtree_4T_notung@data$D, "Y", "D")
iqtree_4T_notung@data$D <- na_if(iqtree_4T_notung@data$D, "N")

Gene_4T_not <- ggtree(iqtree_4T_notung)  %>% rotate(48) %<+% tipnames_4T_notung +
  geom_rootedge(rootedge = 0.25)+
    # geom_text(aes(label=node)) +
  geom_tiplab(aes(label= code_name, color = species), size = 2) +
  scale_color_manual(values=color_palette)+ #unname(glasbey.colors(25)[2:25]) 
  geom_label(aes(label = D), fill = "red", size = 2) +
  guides(color=guide_legend(ncol=1)) +
  theme(legend.title = element_text(size=5),
        legend.text = element_text(size=5),
        legend.position = "left",
        plot.margin = unit(c(0.5,4,3,0), "cm"))+
  coord_cartesian(clip = 'off')

Gene_4T_not

ggsave("4T_notung_tree.pdf", plot = Gene_4T_not,
       device = "pdf", 
       width = 215.9, 
       height =279.4, 
       units = "mm",
       dpi=300)




#final big tree

#now read in the mapped file
iqtree_bs <- read.newick("RAxML bootstrap_2_notaxnames WITH CORRECT MODEL/Cat bootstrap/mapping/RAxML_bipartitions.mapped328_rooted")
tree_annot <- read.csv("pdl_fasta.csv")
tree_annot <- tree_annot %>% 
  mutate(a = row_number()-1) %>%
  mutate(b = "t") %>%
  unite("tip_aln", "b","a", sep = "") %>% 
  relocate(tip_aln) %>%
  rename(tip.label = UP_ID) %>%
  mutate(tip.label = str_replace(tip.label, " ", "_")) %>%
  mutate(tip.label = str_replace(tip.label, "\\:", "_"))%>%
  mutate(tip_a = tip_aln) %>%
  relocate(tip_aln, tip_a)

tree_annot <- tree_annot %>%
  mutate(group = if_else(str_detect(species, "elegans|musculus|cerevisiae"), "Opisthokonta", "Archaeplastida")) %>%
  unite(Text_color_label, species_short, protein_type_, trace_ua, sep = " ", remove = FALSE)




iqtree_annot <- ggtree(iqtree_bs, layout="circular") %<+% tree_annot_withQMATH +
  geom_rootedge(rootedge = 0.25) +
  geom_tiplab(aes(label= new_labels4T, color = simp_type), size = 2) +
  geom_tippoint(aes(fill = domain_size), shape = 21, size = 2) +
  geom_text(aes(label=node)) +
  geom_hilight(node = 762, alpha = 0.15, fill = "aquamarine", extendto = 4) +
    geom_hilight(node = 1321, alpha = 0.15, fill = "limegreen", extendto = 4) +
      geom_hilight(node = 1003, alpha = 0.15, fill = "blue", extendto = 4) +
      geom_hilight(node = 1033, alpha = 0.15, fill = "chartreuse3", extendto = 4) +
      geom_hilight(node = 952, alpha = 0.15, fill = "slateblue4", extendto = 4) +
      geom_hilight(node = 1499, alpha = 0.15, fill = "darkgoldenrod", extendto = 4) +
      geom_hilight(node = 1049, alpha = 0.15, fill = "firebrick", extendto = 4) +
      geom_hilight(node = 1106, alpha = 0.15, fill = "deeppink4", extendto = 4) +

  
  
  geom_text2(aes(label=label, subset = !is.na(as.numeric(label)) & as.numeric(label) > 50, color = !is.na(as.numeric(label)) & as.numeric(label) <75),  size = 2) #hjust=2, vjust =1,

  
# iqtree_annot <- groupClade(iqtree_bs, c(790, 858, 1003, 1033, 952, 1499, 1049, 1106))
#   iqtree_annot <- ggtree(iqtree_bs, aes(color = group), layout =)
# 
# group 790 - TTTT
#   group 858 - 2T
#   group 1003 - ranun specific expansion
#   group 1033 - ranun tp clade
#   group 952 - TP group
#   group 1499 ??
#     group 1049 - 2 T group
#   group 1106 - BTB
#   
  
# raxml_annot_ <- gheatmap(raxml_annot, tree_annot %>% column_to_rownames(var = "tip.label") %>%dplyr::select(species), offset= 1, width = 0.03, colnames = FALSE)
# iqtree_annot

ggsave("IQ-TREE_best_tree_annot_final_one.pdf", plot = iqtree_annot,
       device = "pdf", 
       width = 1000, 
       height =1000, 
       units = "mm",
       dpi=300,
       limitsize = FALSE)





iqtree_annot2 <- ggtree(iqtree_bs) %<+% tree_annot_withQMATH +
  geom_rootedge(rootedge = 0.25) +
  geom_tiplab(aes(label= newlabels_4T, color = species_short), size = 1.5) + #
    scale_color_manual(values=color_palette) +
  # geom_tippoint(aes(fill = trace), shape = 21, size = 2) +
  # geom_text(aes(label=node), size =1) +
  geom_cladelabel(node=762, label="MATH-BTB", align=T, offset = 1, color='black')+
  geom_cladelabel(node=1321, label="MATH-UBP", align=T,  color='black')+
  geom_cladelabel(node=1267, label="Double TRAF, 2nd position", align=T,  color='black')+
  geom_cladelabel(node=1232, label="Double TRAF, 1st position", align=T,  color='black')+  
  geom_cladelabel(node=1417, label="QMATH", align=T,  color='black')+
  geom_cladelabel(node=1482, label="MUSE13/14 (Single TRAF)", align=T,  color='black')+
  geom_cladelabel(node=1167, label="Double TRAF, 1st position", align=T,  color='black')+
  geom_cladelabel(node=1373, label="Papaver MATH-only", align=T,  color='black')+ 
    new_scale_color()+
  geom_nodepoint(aes(label=label, subset = !is.na(as.numeric(label)), color = !is.na(as.numeric(label)) & as.numeric(label) <75),  size = 1.5, stroke =0)+
  scale_color_manual(values=c("black", "red"), guide = FALSE)+
    # geom_text2(aes(label=label, subset = !is.na(as.numeric(label)), color = !is.na(as.numeric(label)) & as.numeric(label) <75),  size = 2)+ #hjust=2, vjust 
    theme(legend.title = element_text(size=5),
        legend.text = element_text(size=5),
        legend.position = c(0.1,0.92),
        plot.margin = unit(c(0.5,2.5,4,0), "cm"))+
  coord_cartesian(clip = 'off')

ggsave("IQ-TREE_best_tree_annot_final_linear.pdf", plot = iqtree_annot2,
       device = "pdf", 
       width = 215.9, 
       height =1000, 
       units = "mm",
       dpi=300)


iqtree_annot2 <- ggtree(iqtree_bs, layout = "circular") %<+% tree_annot_withQMATH +
  geom_rootedge(rootedge = 0.25) +
  geom_tiplab(aes(label= newlabels_4T, color = species_short), size = 1) + #
    scale_color_manual(values=color_palette) +
  # geom_tippoint(aes(fill = trace), shape = 21, size = 2) +
  # geom_text(aes(label=node), size =1) +
  geom_cladelabel(node=762, label="MATH-BTB", align=T, offset = 1, offset.text = 0.25,color='black')+
  geom_cladelabel(node=1321, label="MATH-UBP", align=T,  offset.text = 0.25,hjust = 1, color='black')+
  geom_cladelabel(node=1267, label="Double TRAF, 2nd position", offset.text = 0.25,hjust = 1, align=T,  color='black')+
  geom_cladelabel(node=1232, label="Double TRAF, 1st position", align=T, hjust = 0.5, offset.text = 0.25, color='black')+  
  geom_cladelabel(node=1417, label="QMATH", align=T,  offset.text = 0.25,color='black')+
  geom_cladelabel(node=1482, label="MUSE13/14 (Single TRAF)", align=T,  offset.text = 0.25,color='black')+
  geom_cladelabel(node=1167, label="Double TRAF, 1st position", align=T,  offset.text = 0.25,color='black')+
  geom_cladelabel(node=1373, label="Papaver MATH-only", align=T,  offset.text = 0.25, hjust = 1, color='black')+ 
    new_scale_color()+
  geom_nodepoint(aes(label=label, subset = !is.na(as.numeric(label)), color = !is.na(as.numeric(label)) & as.numeric(label) <75),  size = 1.5, stroke =0)+
  scale_color_manual(values=c("black", "red"), guide = FALSE)+
    # geom_text2(aes(label=label, subset = !is.na(as.numeric(label)), color = !is.na(as.numeric(label)) & as.numeric(label) <75),  size = 2)+ #hjust=2, vjust 
    theme(legend.title = element_text(size=8),
        legend.text = element_text(size=8),
        legend.position = c(0,0.5),
        plot.margin = unit(c(0,0,3,0), "cm"))
  # coord_cartesian(clip = 'off')

ggsave("IQ-TREE_best_tree_annot_final_circular.pdf", plot = iqtree_annot2,
       device = "pdf", 
       width = 450, 
       height =350, 
       units = "mm",
       dpi=300)



#species tree

species_tree <- read.newick("ensembl_all_phyliptree2_new2.phy")
species_tree$tip.label <- str_replace(species_tree$tip.label, "_", " ")



tree_annot_species <- tree_annot_withQMATH %>%
  relocate(species_short)

iqtree_annot3 <- ggtree(species_tree)+ #%<+% tree_annot_species +
  geom_rootedge(rootedge = 0.25) +
  geom_tiplab(aes(label= label, color = label), size = 2) + #
    scale_color_manual(values=color_palette, guide = FALSE) +
    theme(plot.margin = unit(c(0,3,3,0), "cm"))+
  coord_cartesian(clip = 'off')


iqtree_annot3
ggsave("Species_tree.pdf", plot = iqtree_annot3,
       device = "pdf", 
       width = 215.9, 
       height =279.4, 
       units = "mm",
       dpi=300)











tipnames_4T <- tipnames_4T %>%
  mutate(tip.label = Acc)

tree_annot_withQMATH <- left_join(tree_annot,tipnames_4T, by = "uniprot_accession") %>%
  unite(newlabels_4T, species_short, uniprot_accession, sep = "_", remove = FALSE) %>%
  mutate(new_name = coalesce(code_name, newlabels_4T)) %>%
  rename(group_taxa = group) %>%
    mutate(species_short = factor(species_short, levels = species_factor))




iqtree_4T_domain <- tree_subset(iqtree_bs, node=1416, levels_back=0)
iqtree_4T_domain <- groupClade(iqtree_4T_domain, c(98,116,147,132,162))


iqtree_annot_4T_domain <- ggtree(iqtree_4T_domain) %<+% tree_annot_withQMATH + #, aes(color =group)
  geom_rootedge(rootedge = 0.25) +
  # # geom_text(aes(label=node)) +
  # scale_color_manual(values=c("black", "red"), guide = FALSE)
  geom_cladelabel(node=98, label="1st Domain", align=F,  offset = 0.8, color='black')+
  geom_cladelabel(node=116, label="2nd Domain", align=F,  offset = 0.8,color='black')+
  geom_cladelabel(node=147, label="3rd Domain", align=F,  offset = 0.85,color='black')+
  geom_cladelabel(node=132, label="4th Domain", align=F,  offset = 0.9,color='black')+
  geom_cladelabel(node=162, label="Single TRAF", align=F,  offset = 0.95,color='black')+
  new_scale_color()+
  geom_text2(aes(label=label, subset = !is.na(as.numeric(label)), color = !is.na(as.numeric(label)) & as.numeric(label) <75), hjust=1.5, vjust =-0.5, size = 2) +  
  scale_color_manual(values=c("black", "red"), guide = FALSE) +
  new_scale_color() +
  geom_tiplab(aes(label= new_name, color = species_short), size = 2) +
  scale_color_manual(values=color_palette) +
  # geom_tippoint(aes(fill = trace), shape = 21, size = 2) +
  guides(color=guide_legend(ncol=1, )) +
  # xlim(0,5) +
  theme(legend.title = element_text(size=5),
        legend.text = element_text(size=5),
        legend.position = "left",
        plot.margin = unit(c(0.5,2,3,0), "cm"))+
  coord_cartesian(clip = 'off')

iqtree_annot_4T_domain

ggsave("IQ-TREE_best_tree_annot_4T_domain_final_one.pdf", plot = iqtree_annot_4T_domain,
       device = "pdf", 
       width = 215.9, 
       height =279.4, 
       units = "mm",
       dpi=300)


























### OLD
##Now do the small tree

TTTT_group <- tibble(tip_aln = c(tips(iqtree_bs,790))) %>% #, "t134")
  left_join(tree_annot) %>%
  distinct(Gene_ID, .keep_all = TRUE)

write.fasta(as.list(TTTT_group$seq), names= TTTT_group$tip.label, file.out = "June10_TTTT_fullseq.fasta", as.string = TRUE)

raxml_4T_bs <- read.tree("RAxML-2021.06.11-Run.2/combined with boostrap/RAxML_bipartitions.mapped")
raxml_4T_bs <- rotate(raxml_4T_bs, 29)
# raxml_4T_bs <- root.phylo(raxml_4T_bs, 9)
raxml_4T_bs_map <- read.delim("RAxML-2021.06.11-Run.2/taxonNamesTranslationTable.txt", header = FALSE, col.names = c("tip_4T", "tip.label"))
tree_annot_4_TRAF <- left_join(raxml_4T_bs_map,tree_annot) %>%
  unite(tiplab_4T, species_short, uniprot_accession, remove = FALSE, sep = " ")

raxml_4T_bs_annot <- ggtree(raxml_4T_bs) %<+% tree_annot_4_TRAF +
  geom_tiplab(aes(label= tiplab_4T, colour = species_short), size = 2) +
  geom_tippoint(aes(fill = simp_type), shape = 21, size = 2) +
  geom_text(aes(label=node)) # use to find the node use to root
  geom_text2(aes(label=label, subset = !is.na(as.numeric(label)) & as.numeric(label) > 50), hjust=2, vjust =1, size = 2)
  
  
raxml_4T_bs_annot <-  %>% ggtree::rotate(raxml_4T_bs_annot, 25)  
raxml_4T_bs_annot

write.tree(raxml_4T_bs, file = "RAxML-2021.06.11-Run.2/combined with boostrap/RAxML_bipartitions.trial.phy", digits = 15)
raxml_4T_bs <- read.tree("RAxML-2021.06.11-Run.2/combined with boostrap/RAxML_bipartitions.trial.phy")

tree_annot_4_TRAF <- tree_annot_4_TRAF %>%
  group_by(species) %>%
  mutate(Species_Code = substring(species_short, 1, 1)) %>%
  mutate(Species_Code2 = word(species_short, 2,2)) %>%
  mutate(Species_Code2 = substring(Species_Code2, 1, 1))%>%
  mutate(TC1 = "TC1") %>%
  mutate(TRAF4_name = letters[row_number()]) %>%
  unite(TRAF4_name, Species_Code, Species_Code2, TC1, TRAF4_name, sep = "") %>%
  relocate(TRAF4_name, .after = Gene_ID)

tax_TT_4T <- read_tsv("RAxML-2021.06.11-Run.2/taxonNamesTranslationTable.txt", col_names = c("tip_4T_aln", "tip.label")) %>%
  left_join(tree_annot_4_TRAF) %>%
  mutate(species_clean = str_replace_all(species_short, " ","_"))%>%
  unite(TRAF4_name, TRAF4_name, species_clean, sep = "_") %>%
  dplyr::select(tip_4T_aln, TRAF4_name, tip.label)

write_delim(tax_TT_4T, "4T_mapping_for_notung.txt", delim="\t")



notung <- read.tree(text = "(((((((((((PtTC1d_Populustrichocarpa:1.7568263701644562,PtTC1a_Populustrichocarpa:0.630948565559083)r9276,PtTC1b_Populustrichocarpa:1.00000050002909E-6)r9275,AtTC1a_Arabidopsisthaliana:0.03516842599382719)r9274,(PtTC1c_Populustrichocarpa:0.03395433561533473,AtTC1b_Arabidopsisthaliana:0.2761813996383213)r9277)r9273,(SlTC1a_Solanumlycopersicum:0.19387004935067587,DcTC1a_Daucuscarota:0.18975756771783425)r9278)r9272,PsTC1a_Papaversomniferum:0.23155803030455363)r9271,n1086*LOST)r9876,AtTC1a_Amborellatrichopoda:0.3152892932639884)r9269,(SmTC1a_Selaginellamoellendorffii:0.0037479163388582652,SmTC1b_Selaginellamoellendorffii:0.015018177211969743)n875:0.6282835067083631)r9268,(MpTC1a_Marchantiapolymorpha:0.1153313863913965,(((PpTC1b_Physcomitriumpatens:1.00000050002909E-6,PpTC1c_Physcomitriumpatens:0.003215393993995749)r9283,PpTC1d_Physcomitriumpatens:0.008160730663173618)r9282,PpTC1a_Physcomitriumpatens:0.16586279087812805)r9281)r9280)r9267,CbTC1a_Charabraunii:0.38208257355223985)n7900;")

notung_plot <- ggtree(notung) +
  geom_tiplab(aes(label= label), size = 2) +
  # geom_text(aes(label=node)) +
  geom_label(aes(label = c(rep(NA,26),"D",NA,"D","D",NA,NA,"D",NA,"D","D","D")), fill = "red") +
  geom_nodelab(aes(label = c(rep(NA,19),"1.0",NA,"0.8056", "0.587",NA, "0.5278","0.6186", NA, NA, "0.7815", "0.5278", NA, "0.910","0.800","0.700", "0.810","0.880","0.62")), vjust = 1.5) + 
  geom_label(aes(label = c(rep(NA,9),"L",rep(NA,27))), fill = "blue") + 
  ggplot2::xlim(0,15)
  # geom_nodelab(aes(label = c(rep("","26"),"D","","D","D","","","D","","D","D","D")), colour = "red") #node30, 29, 27 33 35 36 37
notung_plot



ggsave("4T_notung.pdf", plot = notung_plot,
       device = "pdf", 
       width = 200, 
       height =100, 
       units = "mm",
       dpi=300,
       limitsize = FALSE)






#for notung

tax_TT <- read_tsv("RAxML-2021.06.08-Run.1/taxonNamesTranslationTable.txt", col_names = c("tip.label", "UP_ID"))
tree_annot <- left_join(tree_annot,tax_TT, by = "UP_ID") %>%
  relocate(tip.label)

notung <- tree_annot %>%
  unite(notung_id, species_short, uniprot_accession, sep = " ", remove = FALSE) %>%
  unite(notung_id, notung_id,Gene_ID,id,n,sep= "", remove = FALSE) %>%
  mutate(notung_id = str_replace_all(notung_id, "_", "")) %>%
  mutate(notung_id = str_replace_all(notung_id, "\\.", "")) %>%
  mutate(notung_id = str_replace_all(notung_id, "\\:", "")) %>%

  dplyr::select(tip.label, species_short)
write_delim(notung, "notung.txt", delim = "\t")






  
```